packages/react-button/src/components/MenuButton/useMenuButton.ts

```diff
@@ -0,0 +1,82 @@
+import * as React from 'react';
+import { getCode, ArrowDownKey } from '@fluentui/keyboard-key';
+import { mergeSlotProp, ComposePreparedOptions } from '@fluentui/react-compose';
+import { useControllableValue, useMergedRefs } from '@uifabric/react-hooks';
+import { DirectionalHint } from 'office-ui-fabric-react';
+import { useButton } from '../Button/useButton';
+import { MenuButtonProps, MenuButtonState } from './MenuButton.types';
+
+/**
+ * The useMenuButton hook processes the MenuButton component props and returns state.
+ * @param props - MenuButton props to derive state from.
+ */
+export const useMenuButton = (
+  props: MenuButtonProps,
+  ref: React.Ref<HTMLElement>,
+  options: ComposePreparedOptions,
+): MenuButtonState => {
+  const {
+    defaultExpanded = false,
+    expanded: controlledExpanded,
+    menu,
+    menuIcon,
+    onClick,
+    onKeyDown,
+    onMenuDismiss,
+    ...rest
+  } = props;
+  const [expanded, setExpanded] = useControllableValue(controlledExpanded, defaultExpanded);
+  const buttonRef = React.useRef(null);
+
+  const _onClick = (ev: React.MouseEvent<HTMLButtonElement>) => {
+    if (onClick) {
+      onClick(ev);
+
+      if (ev.defaultPrevented) {
+        return;
+      }
+    }
+
+    setExpanded(!expanded);
+  };
+
+  const onDismiss = () => {
+    if (onMenuDismiss) {
+      onMenuDismiss();
+    }
+
+    setExpanded(false);
+  };
+
+  const _onKeyDown = (ev: React.KeyboardEvent<HTMLButtonElement>) => {
+    if (onKeyDown) {
+      onKeyDown(ev);
+
+      if (ev.defaultPrevented) {
+        return;
+      }
+    }
+
+    if ((ev.altKey || ev.metaKey) && getCode(ev) === ArrowDownKey) {
+      setExpanded(true);
+    }
+  };
+
+  const state = {
+    ...rest,
+    'aria-expanded': expanded,
+    expanded,
+    onClick: _onClick,
+    onKeyDown: _onKeyDown,
+    ref: useMergedRefs(ref, buttonRef),
+
+    // Menu slot props.
+    menu: mergeSlotProp(menu, {
+      directionalHint: DirectionalHint.bottomRightEdge,
+      onDismiss,
+      target: buttonRef && buttonRef.current,
+    }),
+  };
+
+  return useButton(state, ref, options) as MenuButtonState;
```

Comment:
i think instead of `as MenuButtonState` here. we should fix typing of `useButton` to be similar to what you had for `useToggle`

Reply:
I think it is a different scenario since `useToggle` is supposed to be a more generic thing that could be reusable across different components while `useButton` is very much linked to the `Button` component. However, I'm happy to discuss this if need be.

Reply:
i don't quite follow. where i am getting at is we should not need to do `as MenuButtonState`

Reply:
I've improved the typings so that we don't need this cast.